learningequality/studio
# Handling copy failures and retry capability

## Summary

This PR brings in the capability to retry failed copies and detection of failing copies.

### Manual verification steps performed

To generate copy error state on the frontend, I raised an exception from the backend copy function. These are the checks I've performed:-
 
1. Checked whether error in the `changesForSyncing` table are detected. (happens when copy fails due to a known reason)
2. Checked whether task errors are detected. (happens due to abnormal task termination or unhandled exceptions)
3. Checked whether importing from channels errors out.
4. Checked whether copying completes.
5. Checked whether retrying completes successfully when the exception is removed from the backend code.
6. Checked whether after multiple failures, retrying still reaches success upon exception removal.
7. Copying multiple nodes together by selecting them works as expected.
8. When copy is failed, the node can be removed.

### Screenshots (if applicable)

The below screenshot is a work in progress. I need to make some CSS changes to make the UI look as it is expected.

![image](https://user-images.githubusercontent.com/26724128/236506819-ab898b0f-6cd5-4c99-a9c1-7228bfa4cc4e.png)

The below is the UI design that is expected.

![image](https://user-images.githubusercontent.com/6668144/110513952-a2975b00-80bb-11eb-9dda-4c078facce96.png)


### Does this introduce any tech-debt items?

I tested the code thoroughly and I've the following concerns, they might be or might not be tech debts depending on how we treat them, but listing them out for clarity:-

1. Due to the architectural design of this PR, signing out and signing in doesn't persist the error state.
2. The error state doesn't get synced between multiple users. 

___
## Reviewer guidance




## References

Closes https://github.com/learningequality/studio/issues/2850.
Closes https://github.com/learningequality/studio/issues/3914.

----

### Contributor's Checklist


PR process:


Studio-specifc:



Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
